Fix PAC watcher panic when concurrency_limit is set

zakisk opened the pull request:
This fixes a panic in PAC watcher when concurreny_limit is set via global repository and concurrency was not working as expected.

Solves these two issue:
https://issues.redhat.com/browse/SRVKP-5560

enarha commented on pkg/reconciler/queue_pipelineruns.go:
```diff
@@ -34,6 +34,19 @@ func (r *Reconciler) queuePipelineRun(ctx context.Context, logger *zap.SugaredLo
+		repo.Spec.Merge(globalRepo.Spec)
+	}
+
+	if repo.Spec.GitProvider != nil && repo.Spec.GitProvider.Secret == nil && r.globalRepo.Spec.GitProvider != nil && r.globalRepo.Spec.GitProvider.Secret != nil {
```
Shouldn't we also check that the GitProvider on the "local" repo and global repo match? The documentation says "Since the local repository CR has the git_provider.type gitlab like the global repository CR the git provider settings for the GitLab will be taken from the global repository. The secret referenced will be fetched from where the global repository is defined." . The way I read it, they should match to take the secret from the global repo config.

zakisk replied:
@enarha I got answer to your this question now :smile: that we [merge](https://github.com/openshift-pipelines/pipelines-as-code/blob/d56bd5e4bc242719431e41c644ca2b295676f8a5/pkg/apis/pipelinesascode/v1alpha1/types.go#L148) repo with global (you can see above) in that merge it is confirmed that both git providers are equal.

enarha commented on test/gitea_test.go:
```diff
@@ -720,6 +720,110 @@ func verifyProvinance(t *testing.T, topts *tgitea.TestOpts, expectedOutput, cNam
+	topts.CheckForStatus = "success"
+	tgitea.WaitForStatus(t, topts, "heads/"+targetRef, "", false)
+	// check the output of the PipelineRun logs
+	err = twait.RegexpMatchingInPodLog(context.Background(), topts.ParamsRun, topts.TargetNS, "pipelinesascode.tekton.dev/event-type=pull_request", cName, *regexp.MustCompile(expectedOutput), "", 2)
```
I do not understand this test. There are 5 pipelineruns defined, right? With the concurrency limit of 2, I expect all 5 of them to run, but only 2 of them concurrently. Eventually you should get here 5 matches, but you only get 2.

zakisk replied:
@enarha got you, I am changin it.

enarha replied:
Can you elaborate how are you trying to change it? I mean describe in words what's the purpose of this test and how you want to implement it?

zakisk replied:
I want to test that `concurrency_limit` is working while set by global repository and in other test `concurrency_limit`.

zakisk replied:
previously I copied all the code from previous tests and messed it, now I will be pushing updated clean code.

enarha replied:
I still do not think we actually verify the concurrency here. To verify the concurrency we need to verify that no more than 2 (in this case) pipelineruns are running concurrently. I do understand the test as is, do check the case where the bug was encountered and fixed. Bottom line is I think we can merge that PR as is (given the failing test pass), but we need to ensure we have a test which covers the actual concurrency as well (maybe we do? I'm still not familiar with the code in the repo.).

zakisk replied:
yes @enarha, I agree with you it should check for concurrent PR and it needs to figure out a way that how to do it because at the moment other already written PAC tests do it in the same way I did it.

enarha replied:
OK. As I mentioned, we can merge that as is and I'm resolving this thread.

savitaashture commented:
@zakisk i m fixing go vet issues in this PR https://github.com/openshift-pipelines/pipelines-as-code/pull/1749

piyush-garg commented on pkg/reconciler/queue_pipelineruns.go:
```diff
@@ -34,6 +34,13 @@ func (r *Reconciler) queuePipelineRun(ctx context.Context, logger *zap.SugaredLo
 	}
 
+	// find global repository if set
+	globalRepo, err := r.repoLister.Repositories(r.run.Info.Kube.Namespace).Get(r.run.Info.Controller.GlobalRepository)
```
do we need to fetch here, it is in reconciler right?

zakisk replied:
yeah, it is reconciler. I will change that.

savitaashture commented on pkg/reconciler/queue_pipelineruns.go:
```diff
@@ -34,6 +34,10 @@ func (r *Reconciler) queuePipelineRun(ctx context.Context, logger *zap.SugaredLo
 		return fmt.Errorf("updateError: %w", err)
 	}
 
+	if r.globalRepo, err = r.repoLister.Repositories(r.run.Info.Kube.Namespace).Get(r.run.Info.Controller.GlobalRepository); err == nil && r.globalRepo != nil {
```
Can you add a comment for this change

savitaashture commented on test/github_pullrequest_concurrency_test.go:
```diff
@@ -46,14 +53,49 @@ func TestGithubSecondPullRequestConcurrency1by1WithError(t *testing.T) {
+	label := "Github PullRequest Concurrent Two at a Time Set by Global Repo"
+	numberOfPipelineRuns := 10
+	// in this test we don't want to have `concurrency_limit` in local repo because
+	// global repository's `concurrency_limit` is being tested here.
```
```suggestion
	// In this test, we do not set `concurrency_limit` in the local repository
        // as the goal is to verify the `concurrency_limit` for the global repository.
```

savitaashture commented on test/github_pullrequest_concurrency_test.go:
```diff
@@ -49,6 +56,40 @@ func TestGithubSecondPullRequestConcurrency1by1WithError(t *testing.T) {
+	numberOfPipelineRuns := 10
+	// in this test we don't want to have `concurrency_limit` in local repo because
+	// global repository's `concurrency_limit` is being tested here.
+	maxNumberOfConcurrentPipelineRuns := -1
```
Can we rename to different name as its meaning is confusing when concurrency is set by globalRepo

How about 

`localRepoMaxConcurrentRuns` or `maxConcurrentPipelineRunsLocalRepo`

savitaashture commented on test/github_pullrequest_concurrency_test.go:
```diff
@@ -155,3 +198,44 @@ func testGithubConcurrency(ctx context.Context, t *testing.T, maxNumberOfConcurr
+	err := runcnx.Clients.NewClients(ctx, &runcnx.Info)
+	if err != nil {
+		return ctx, "", nil, err
+	}
```
```suggestion
	if err := runcnx.Clients.NewClients(ctx, &runcnx.Info); err != nil {
		return ctx, "", nil, err
	}
```

savitaashture commented on test/github_pullrequest_concurrency_test.go:
```diff
@@ -155,3 +198,44 @@ func testGithubConcurrency(ctx context.Context, t *testing.T, maxNumberOfConcurr
+	err = repository.CreateRepo(ctx, globalNS, runcnx, repo)
+	if err != nil {
+		return ctx, "", nil, err
+	}
```
```suggestion
	if err = repository.CreateRepo(ctx, globalNS, runcnx, repo); err != nil {
		return ctx, "", nil, err
	}
```

savitaashture commented on test/github_pullrequest_concurrency_test.go:
```diff
@@ -155,3 +198,44 @@ func testGithubConcurrency(ctx context.Context, t *testing.T, maxNumberOfConcurr
+		runcnx.Clients.Log.Infof("Cleaning up global repo %s in %s", info.DefaultGlobalRepoName, globalNS)
+		err := runcnx.Clients.PipelineAsCode.PipelinesascodeV1alpha1().Repositories(globalNS).Delete(
+			context.Background(), info.DefaultGlobalRepoName, metav1.DeleteOptions{})
+		return err
```
```suggestion
		return runcnx.Clients.PipelineAsCode.PipelinesascodeV1alpha1().Repositories(globalNS).Delete(
			context.Background(), info.DefaultGlobalRepoName, metav1.DeleteOptions{})
```

enarha commented:
/lgtm

savitaashture commented on test/github_pullrequest_concurrency_test.go:
```diff
@@ -49,6 +56,40 @@ func TestGithubSecondPullRequestConcurrency1by1WithError(t *testing.T) {
+
+	numberOfPipelineRuns := 10
+	maxNumberOfConcurrentPipelineRuns := 3
+	testGithubConcurrency(ctx, t, maxNumberOfConcurrentPipelineRuns, numberOfPipelineRuns, label, false, map[string]string{})
```
Can you move this to different function and call that function from this test and TestGithubGlobalRepoConcurrencyLimit test to avoid duplication

Something like below

```
func TestGithubGlobalRepoConcurrencyLimit(t *testing.T) {
         testGlobalRepoConcurrency(t, "Github PullRequest Concurrent Two at a Time Set by Global Repo", -1)
}

func TestGithubGlobalAndLocalRepoConcurrencyLimit(t *testing.T) {
         testGlobalRepoConcurrency(t, "Github PullRequest Concurrent Three at a Time Set by Local Repo", 3)
}

func testGlobalRepoConcurrency(t *testing.T, label string, maxConcurrentPipelineRunsLocalRepo int) {
        ctx := context.Background()
	// create global repo
	ctx, globalNS, runcnx, err := createGlobalRepo(ctx)
	assert.NilError(t, err)
	defer (func() {
		err = cleanUpGlobalRepo(runcnx, globalNS)
		assert.NilError(t, err)
	})()

	numberOfPipelineRuns := 10
	testGithubConcurrency(ctx, t, maxConcurrentPipelineRunsLocalRepo, numberOfPipelineRuns, label, false, map[string]string{})
}
```

Commit:
Fix PAC watcher panic when concurrency_limit is set

This fixes a panic in PAC watcher when concurreny_limit is
set via global repository and concurrency was not working as
expected. E2E tests are also added.

https://issues.redhat.com/browse/SRVKP-5560

Signed-off-by: Zaki Shaikh <zashaikh@zashaikh-thinkpadp1gen4i.bengluru.csb>

savitaashture commented:
/lgtm

cc @piyush-garg

piyush-garg commented on pkg/reconciler/queue_pipelineruns.go:
```diff
@@ -34,6 +34,12 @@ func (r *Reconciler) queuePipelineRun(ctx context.Context, logger *zap.SugaredLo
 
+	// merge local repo with global repo here in order to derive settings from global
+	// for further concurrency and other operations.
+	if r.globalRepo, err = r.repoLister.Repositories(r.run.Info.Kube.Namespace).Get(r.run.Info.Controller.GlobalRepository); err == nil && r.globalRepo != nil {
```
do we need to do listing here? global repo is available in not available in reconciler object?

zakisk replied:
@piyush-garg I confirmed that we need to fetch it, global repo wasn't there before `queue_pipelineruns`.

piyush-garg replied:
ack, got it

piyush-garg commented:
/lgtm
